packages/rpc/src/vatidation/BundlerCollectorTracer.ts

```diff
@@ -0,0 +1,365 @@
+            if (
+                log.getGas() < log.getCost() ||
+                // special rule for SSTORE with gas metering
+                (opcode === "SSTORE" && log.getGas() < 2300)
+            ) {
+                this.currentLevel.oog = true
+            }
+
+            if (opcode === "REVERT" || opcode === "RETURN") {
+                if (log.getDepth() === 1) {
+                    // exit() is not called on top-level return/revent, so we reconstruct it
+                    // from opcode
+                    const ofs = parseInt(log.stack.peek(0).toString())
+                    const len = parseInt(log.stack.peek(1).toString())
+                    const data = toHex(log.memory.slice(ofs, ofs + len)).slice(0, 4000) as Hex
+                    // this.debug.push(opcode + ' ' + data)
+                    this.calls.push({
+                        type: opcode,
+                        gasUsed: 0,
+                        data
+                    })
+                }
+                // NOTE: flushing all history after RETURN
+                this.lastThreeOpcodes = []
+            }
+
+            if (log.getDepth() === 1) {
+                if (opcode === "CALL" || opcode === "STATICCALL") {
+                    // stack.peek(0) - gas
+                    const addr = toAddress(log.stack.peek(1).toString(16))
+                    const topLevelTargetAddress = toHex(addr)
+                    // stack.peek(2) - value
+                    const ofs = parseInt(log.stack.peek(3).toString())
+                    // stack.peek(4) - len
+                    const topLevelMethodSig = toHex(log.memory.slice(ofs, ofs + 4))
+
+                    this.currentLevel = this.callsFromEntryPoint[this.topLevelCallCounter] = {
+                        topLevelMethodSig,
+                        topLevelTargetAddress,
+                        access: {},
+                        opcodes: {},
+                        extCodeAccessInfo: {},
+                        contractSize: {}
+                    }
+                    this.topLevelCallCounter++
+                } else if (opcode === "LOG1") {
+                    // ignore log data ofs, len
+                    const topic = log.stack.peek(2).toString(16)
+                    if (topic === this.stopCollectingTopic) {
+                        this.stopCollecting = true
+                    }
+                }
+                this.lastOp = ""
+                return
+            }
+
+            const lastOpInfo = this.lastThreeOpcodes[this.lastThreeOpcodes.length - 2]
+            // store all addresses touched by EXTCODE* opcodes
+            if (lastOpInfo && lastOpInfo.opcode && lastOpInfo.opcode.match(/^(EXT.*)$/) != null) {
+                const addr = toAddress(lastOpInfo.stackTop3[0].toString(16))
+                const addrHex = toHex(addr)
+                const last3opcodesString = this.lastThreeOpcodes.map((x) => x.opcode).join(" ")
```

Comment:
why do we convert this to a string and use regex logic instead of just working with the array?

Reply:
Done